File: 678-valid-parenthesis-string/valid-parenthesis-string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        mi_open = 0
        ma_open = 0
        for i in s:
            if(i=='('):
                mi_open+=1
                ma_open+=1
            elif(i==')'):
                mi_open-=1
                ma_open-=1
            else:
                mi_open-=1
                ma_open+=1
            
            if(ma_open<0):
                return False
            mi_open = max(mi_open,0)
        return mi_open == 0

File: 678-valid-parenthesis-string/valid-parenthesis-string.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int mi_open = 0;
        int ma_open = 0;
        for(auto i:s){
            if(i=='('){
                mi_open++;
                ma_open++;
            }
            else if(i==')'){
                mi_open--;
                ma_open--;
            }
            else{
                mi_open--;
                ma_open++;
            }
            if(ma_open<0){
                return false;
            }
            mi_open = max(mi_open,0);
        }
        return mi_open == 0;
    }
};